ul/src/pdu/reader.rs

```diff
@@ -0,0 +1,770 @@
+                            codec
+                                .decode(&read_n(&mut cursor, item_length as usize)?)?
+                                .trim()
+                                .to_string(),
+                        );
+                    }
+                    0x40 => {
+                        // Transfer Syntax Sub-Item Structure
+
+                        // 5-xxx - Transfer-syntax-name(s) - This variable field shall contain the
+                        // Transfer-syntax-name proposed for this presentation context. A valid
+                        // Transfer-syntax-name shall be encoded as defined in Annex F. For a
+                        // description of the use of this field see Section 7.1.1.13.
+                        // Transfer-syntax-names are structured as UIDs as defined in PS3.5 (see
+                        // Annex B for an overview of this concept). DICOM Transfer-syntax-names are
+                        // registered in PS3.5.
+                        transfer_syntaxes.push(
+                            codec
+                                .decode(&read_n(&mut cursor, item_length as usize)?)?
+                                .trim()
+                                .to_string(),
+                        );
+                    }
+                    _ => {
+                        return Err(Error::UnknownPresentationContextSubItem);
+                    }
+                }
+            }
+
+            Ok(PduVariableItem::PresentationContextProposed(
+                PresentationContextProposed {
+                    id: presentation_context_id,
+                    abstract_syntax: abstract_syntax.ok_or(Error::MissingAbstractSyntax)?,
+                    transfer_syntaxes,
+                },
+            ))
+        }
+        0x21 => {
+            // Presentation Context Item Structure (proposed)
+
+            let mut transfer_syntax: Option<String> = None;
+
+            // 5 - Presentation-context-ID - Presentation-context-ID values shall be odd integers
+            // between 1 and 255, encoded as an unsigned binary number. For a complete description
+            // of the use of this field see Section 7.1.1.13.
+            let presentation_context_id = cursor.read_u8()?;
+
+            // 6 - Reserved - This reserved field shall be sent with a value 00H but not tested to
+            // this value when received.
+            cursor.read_u8()?;
+
+            // 7 - Result/Reason - This Result/Reason field shall contain an integer value encoded
+            // as an unsigned binary number. One of the following values shall be used:
+            //   0 - acceptance
+            //   1 - user-rejection
+            //   2 - no-reason (provider rejection)
+            //   3 - abstract-syntax-not-supported (provider rejection)
+            //   4 - transfer-syntaxes-not-supported (provider rejection)
+
+            let reason;
+            match cursor.read_u8()? {
```

Comment:
Part of this logic could be moved into an associated function of `PresentationContextResultReason` (something like `fn from_code(code: u8) -> Option<Self>`. And the same could probably be done for other types. I also happen to know that [`num-derive`](https://crates.io/crates/num-derive) lets you derive an implementation of `FromPrimitive` for enum types.